underscore.mdx

```diff
@@ -5,133 +5,146 @@ description: Using underscore expressions to define features
 
-Consider the feature class `RocketEngine` defined below:
+Underscore expressions are useful because Chalk statically analyzes and optimizes them, leading to better performance
+when compared to equivalent Python resolvers. They are also useful for succinctly defining common features, such as the
+number of times a user failed a login attempt over the past 30 days or the total amount spent at a given merchant.
 
-```python
-from chalk.features import features
+## Arithmetic
 
-@features
-class RocketEngine:
-    id: int
-    mass: float
-    volume: float
-```
-
-Suppose we wanted to define a feature called `density` and resolve it as `mass / volume`.
-We can use the underscore feature definition syntax as follows:
+In this example, we have a Transaction feature class with `total` and `sales_tax` features and we want to define
+`subtotal` as `total` minus `sales_tax`. Instead of writing a Python resolver, we can resolve `subtotal` with an
+underscore expression:
 
 ```diff-py
-- from chalk.features import features
+- from chalk import online
 + from chalk.features import _, features
 
 @features
-class RocketEngine:
+class Transaction:
     id: int
-    mass: float
-    volume: float
+    total: float
+    sales_tax: float
++   subtotal: float = _.total - _.sales_tax
 
-+   # Defining density using an underscore expression
-+   density: float = _.mass / _.volume
+- @online
+- def get_subtotal(total: Transaction.total, sales_tax: Transaction.sales_tax) -> Transaction.subtotal:
+-     return total - sales_tax
 ```
 
-Note that `_` is a reference to the containing feature class `RocketEngine` and should
-be used to reference features contained in the same feature class.
-
----
-
-## DataFrames
+## Conditions and filters
 
-Underscore can also be used to define expressions with [`DataFrame`](/docs/dataframe) features.
+DataFrame features can be [filtered](/docs/dataframe#filters) with underscore expressions.
 
-In the example below, we will define two new feature classes, `Astronaut` and `RocketShip`:
+Extending our `Transaction` example, we can create `User` feature class with a [has-many](/docs/has-many) relationship
+to `Transaction`. Then, we can define a feature representing large purchases by referencing the existing
+`User.transactions` feature:
 
-```python
-from chalk.features import features, has_one, has_many, DataFrame
-
-@features
-class Astronaut:
-    id: int
-    name: str
-    height: float
-    mass: float
-    ship_id: int
+```diff-py
+from chalk.features import _, features, DataFrame
 
 @features
-class RocketShip:
+class Transaction:
     id: int
-    engine_id: int
++   user_id: "User.id"
+    total: float
+    sales_tax: float
+    subtotal: float = _.total - _.sales_tax
+
++ @features
++ class User:
++    id: int
++    # implicit has-many relationship with Transaction due to `user_id` above
++    transactions: DataFrame[Transaction]
++    large_transactions: DataFrame[Transaction] = _.transactions[_.total > 1000]
```

Comment:
This is currently unsupported - although you can _use_ dataframe expressions inside of underscores, the actual output always has to be a scalar (with the exception of the materialized aggregate features which are a special case).

So for example, you can say

```
large_transaction_count: int = _.transactions[_.total > 1000].count()
```

or

```
large_transaction_total: int = _.transactions[_.total, _.total > 1000].sum()
```

Reply:
Eventually I would like to be able to lift this restriction but it is not super easy